api/orders.js

```diff
@@ -1,4 +1,232 @@
+ordersRouter.get(
+  "/status/:status",
+  requireUser,
+  requireAdmin,
+  async (req, res, next) => {
+    const { status } = req.params;
+    try {
+      const orders = await getOrdersByStatus(status);
+      res.send(orders);
+    } catch (error) {
+      console.error(error);
+      next({
+        name: "orderDoesNotExist",
+        message: "There are no orders matching status",
+      });
+    }
+  }
+);
+
+// Create a new order on visit and when there is no existing cart. Create a new order when the order status changes to success.
+// Order default status is order_pending.
+// endpoint "/cart"
+ordersRouter.post("/", async (req, res, next) => {
+  const { email, address } = req.body;
+  try {
+    if (!req.user) {
+      const newOrder = await createOrder({
+        userId: 1,
+        email,
+        address,
+      });
+      res.send(newOrder);
+    } else if (req.user) {
+      const { id } = req.user;
+      const newOrder = await createOrder({
+        userId: id,
+        email,
+        address,
+      });
+      res.send(newOrder);
+    }
+  } catch (error) {
+    console.error(error);
+    next({
+      name: "CreateOrderError",
+      message: "Failed to process the order",
+    });
+  }
+});
+
+// endpoint "/checkout". Order status changes to "payment_pending". During this step, users should fill in their payment method such as billing address, credit card info, etc."
+ordersRouter.patch("/checkout", async (req, res, next) => {
+  try {
+    const { id } = req.body;
+    const orderStatus = await setOrderAsPaymentPending(id);
+    res.send(orderStatus);
+  } catch (error) {
+    console.error(error);
+    next({
+      name: "OrderStatusProcessingError",
+      message: "Failed to update the order as processing",
+    });
+  }
+});
+
+// endpoint "/pay" once user confirms pay. Order status changes to "processing."
+ordersRouter.patch("/pay", async (req, res, next) => {
+  try {
+    const { id } = req.body;
+    const orderStatus = await setOrderAsProcessing(id);
+    res.send(orderStatus);
+  } catch (error) {
+    console.error(error);
+    next({
+      name: "OrderStatusSuccessError",
+      message: "Failed to update the order as success",
+    });
+  }
+});
+
+// endpoint "/confirm". Order status changes to "success" once admin manually confirms the order's payment.
+ordersRouter.patch(
+  "/confirm",
+  requireUser,
+  requireAdmin,
+  async (req, res, next) => {
+    try {
+      const { id } = req.body;
+      const orderStatus = await setOrderAsSuccess(id);
+      res.send(orderStatus);
+    } catch (error) {
+      console.error(error);
+      next({
+        name: "OrderStatusSuccessError",
+        message: "Failed to update the order as success",
+      });
+    }
+  }
+);
+
+// This is to update any order info such as email and address. This is unrelated to status updates.
+ordersRouter.patch("/update/:orderId", async (req, res, next) => {
```

Comment:
doesn't follow RESTful api standard

```suggestion
ordersRouter.patch("/:orderId", async (req, res, next) => {
```

Reply:
maybe add requireUser

Reply:
corrected